cli/lib/webpack/service.js

```diff
@@ -0,0 +1,226 @@
+		if (this.webpackStatus.state === 'ready') {
+			this.logger.info('Build is up-to-date and ready!');
+			if (!this.shouldWatch) {
+				this.client.disconnect();
+			}
+			return Promise.resolve();
+		}
+
+		this.logger.info('Waiting for compilation to complete ...');
+		await this.subcribeToWebpackStatusChanges();
+
+		return new Promise((resolve, reject) => {
+			const handler = e => {
+				if (e.state === 'ready' || e.state === 'error') {
+					this.webpackStatusEmitter.off('status', handler);
+
+					if (e.state === 'ready') {
+						clearTimeout(buildTimeout);
+						if (!this.shouldWatch) {
+							this.client.disconnect();
+						}
+						return resolve();
+					} else {
+						return reject(new Error('Webpack compilation failed.'));
+					}
+				}
+			};
+			this.webpackStatusEmitter.on('status', handler);
+
+			// @todo what to do in this case? Give useful tips!
+			const buildTimeout = setTimeout(() => reject(new Error('Did not receive success message from Webpack appcd plugin within 120s, aborting...')), 120000);
+		});
+	}
+
+	startWebpack(options) {
+		this.logger.debug(`Starting Webpack with options: ${JSON.stringify(options)}`);
+		this.webpackStatus = { state: 'building' };
+		return new Promise((resolve, reject) => {
+			this.client
+				.request({
+					path: `${this.basePath}/start`,
+					data: options
+				})
+				.once('response', resolve)
+				.once('error', reject);
+		});
+	}
+
+	subcribeToWebpackStatusChanges() {
+		return new Promise((resolve, reject) => {
+			this.client
+				.request({
+					path: `${this.basePath}/status/${this.jobIdentifier}`,
+					type: 'subscribe'
+				})
+				.on('response', (data, response) => {
+					if (typeof data === 'string' && data === 'Subscribed') {
+						this.sid = response.sid;
+						return resolve();
+					}
+
+					if (data.event === 'state') {
+						this.logger.debug(`Status changed from ${this.webpackStatus.state.cyan} to ${data.state.cyan}`);
+						this.webpackStatus = data;
+						this.webpackStatusEmitter.emit('status', data);
+					} else if (data.event === 'output') {
+						data.output
+							.replace(/\s+$/, '')
+							.split('\n')
+							.forEach(m => this.logger.debug(m));
+					} else if (data.event === 'progress') {
+						this.reporter.progress(data.progress);
+					} else if (data.event === 'done') {
+						this.reporter.done(data);
+					} else if (data.event === 'api-usage') {
+						this.builder.tiSymbols = data.tiSymbols;
+					}
+				})
+				.once('error', err => {
+					reject(err);
+				});
+		});
+	}
+
+	generateJobIdentifier() {
+		const hash = crypto.createHash('sha1');
+		hash.update(this.projectSettings.path);
+		return hash.digest('hex');
+	}
+
+	validateOptions(currentOptions, newOptions) {
+		const checkProperties = [
+			'project',
+			'build'
+		];
+		const pluck = (obj, props) => {
+			return props.reduce((acc, prop) => {
+				acc[prop] = obj[prop];
+				return acc;
+			}, {});
+		};
+		try {
+			assert.deepStrictEqual(
```

Comment:
to avoid the need to try/catch you may want to switch to `util.isDeepStrictEqual()` which returns true/false rather than throws
https://nodejs.org/api/util.html#util_util_isdeepstrictequal_val1_val2

Reply:
Yes, that's much better, thanks!